internal/topology/variables/clusterclass_variable_validation.go

```diff
@@ -95,6 +98,21 @@ func validateClusterClassVariableName(variableName string, fldPath *field.Path)
 	return allErrs
 }
 
+// validateRequiredVariablesAreDefaulted validates that ClusterClass variable names are unique.
+func validateRequiredVariablesAreDefaulted(clusterClassVariable *clusterv1.ClusterClassVariable, pathPrefix *field.Path) field.ErrorList {
```

Comment:
I don't think we should enforce that **all** required variables **must** have default values.

This only applies to variables newly added during a ClusterClass update, because the corresponding Clusters cannot have values for those variables.

In case of a new ClusterClass default values are not required. I think there are cases of required variables where there is simply no valid default value (e.g. azure subscription id: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/cluster-template.yaml#L44)

Reply:
I think we should make this part of our regular "does the new schema match on all Clusters" validation code. This way new required variables can be introduced like this by the user:

* add new variable with required false (aka optional) to the ClusterClass (existence of default value does not matter)
* add variable values to all Clusters
* make the variable required in the ClusterClass

With this kind of validation we can even account for changes between required true/false without any additional code for that special case.

Reply:
+1 to make this part of our regular "does the new schema match on all Clusters" validation code.

Reply:
Can be resolved.